covid19-dashboard_NEW/calculate_data.py

```diff
@@ -0,0 +1,173 @@
+
+    return output
+
+
+def _clean_dataset(date_to_val: dict,
+                   step_size: int = 2,
+                   start_date: str = '2020-03-01'):
+    """
+    Given a dataset, return the elements whose index is divisible by step_size.
+    Remove any negative values and dates starting prior to start_date.
+    :param start_date: any date before this date will be omitted.
+    :param date_to_val: an object of type date->int.
+    :param step_size: the size of our steps.
+    :return: the same dataset but 1/nth of the size.
+    """
+    # The latest date is the largest date from the list of dates.
+    latest_date: str = max(list(date_to_val.keys()))
+
+    # Get a list of (Key, Value).
+    # Key = date, Value = cases/deaths.
+    date_to_val_list = list(date_to_val.items())
+    # Sort the values by date, elem[0].
+    date_to_val_list = sorted(date_to_val_list, key=lambda e: e[0])
+
+    # Only include the element if its index is divisible by chunk.
+    date_to_val_list = date_to_val_list[::step_size]
+
+    # Remove dates before March 1st.
+    date_to_val_list = [
+        (date, value)
+        for date, value in date_to_val_list
+        if date >= start_date
+    ]
+
+    # Remove negative values from the list.
+    date_to_val_list = [
+        (date, 0) if value < 0 else (date, value)
+        for date, value in date_to_val_list
+    ]
+
+    # Because the latest date might have been skipped by the step_size.
+    # Check to see if it's in our dataset, otherwise append it.
+    # We need the latest date to be present.
+    if date_to_val_list[-1][0] != latest_date:
+        date_to_val_list.append((latest_date, date_to_val[latest_date]))
+
+    # Convert the list of (date, value) back to a dictionary.
+    output = dict(date_to_val_list)
+
+    return output
+
+
+def calculate_data(data_type: str, data: dict, population: int) -> dict:
+    """
+    Given a dataset for a region and its population,
+    perform different types of calculations on every date.
+    :param data_type: type of data, "cases" or "deaths".
+    :param data: a dictionary containing date->int.
+    :param population: geoId -> population.
+    :return: a dictionary with the 7-week moving average + other calculations.
+    """
+    # For example, 1-week moving average.
+    chunk_size: int = 7
+
+    # Instantiate the variables to {} or None depending on data type.
+    week_ave: dict = {}
+    week_ave_value: Union[float, None] = None
+    per_capita: dict = {}
+    per_capita_value: Union[float, None] = None
+    pct_increase_value: Union[float, None] = None
+
+    # If there are less elements than our chunk_size.
+    # We can't perform any calculations.
+    if len(data.keys()) > chunk_size:
+        # The latest date is the largest date from the list of dates.
+        latest_date: str = max(list(data.keys()))
+
+        # Calculate the time-series moving average.
+        week_ave: dict = _moving_average(data, chunk=chunk_size)
```

Comment:
consider using names that is more different here. week_ave could read as a single value -- maybe week_moving_averages?

Reply:
Fixed.